---
title: "LIMMA"
author: "DJL"
date: "03/11/2016"
output:
  html_document:
    toc: yes
    toc_float: yes
---
#Overview

**Aim**  
Differential Expression on GAP FEP Data.  


#Pseudocode
```{r Pseudocode, eval=FALSE}

#1.Load Libs

#2.Set Dir

#3.Load Lumibatch

#4. LIMMA
Extract pData
Prepare for Limma (age ethnicity gender Phenotype)
perform limma

#5. Write CSV files
All Probes
Significantly Diff expressed Probes
Significantly Diff expressed Probes + log2 cut off
Non Significant Probes

#6. Volcano Plot




```

## Load Libaries
```{r load_libs, tidy=TRUE}
library(lumi)
library(limma)
library(dplyr)

```

## Set directories
```{r set directories, eval=TRUE, tidy=TRUE}
data_dir <-"./data/"
P1_output_dir <-"./P1_Diff_Ex/output/"
P1_figs_dir <-"./P1_Diff_Ex/figs/"
```



## Load eset
```{r load_eset, tidy=TRUE}
#Load Gene expression data in Lumi batch object

load(paste(data_dir,"GAP_FEP_Full_Gene_Expression_Data.RData",sep=""))

```

## create design matrix
```{r LIMMA Design}

#Prepare covariates
names(pData(eset_bg_log2_rsn_SVA_Good))
gap <- factor(eset_bg_log2_rsn_SVA_Good$Phenotype)
gap_gender <- factor(eset_bg_log2_rsn_SVA_Good$Gender)
gap_age <-as.numeric(eset_bg_log2_rsn_SVA_Good$Age)
gap_ethnicity <-factor(eset_bg_log2_rsn_SVA_Good$Ethnicity)


design = model.matrix(~0+gap+gap_gender+gap_ethnicity+gap_age)  

```


## Limma lmFit
```{r limma_lmFit, tidy=TRUE}

fit <- lmFit(eset_bg_log2_rsn_SVA_Good, design)


```

## Limma makeContrasts 

```{r limma_makeContrasts, tidy=TRUE}
contrasts <- makeContrasts(gapFEP-gapcontrol, levels=design)
```



## Limma eBayes on makeContrasts

```{r limma_eBayes_on_makeContrasts, tidy=TRUE}
contrast.fit1 <- contrasts.fit(fit, contrasts)
contrast.fit1 <- eBayes(contrast.fit1)


```

```{r limma_topTable, comment=NA, width=100, tidy=TRUE}
# Limma topTabl
top_de_genes1 <- topTable(contrast.fit1, coef=1, number=10000,adjust.method="fdr",p.value=1,confint=TRUE)

names(top_de_genes1)

# tidy up
top_de_genes1$logFC <- round(top_de_genes1$logFC,2)
top_de_genes1$CI.L <- round(top_de_genes1$CI.L,2)
top_de_genes1$CI.R <- round(top_de_genes1$CI.R,2)
top_de_genes1$AveExpr <- round(top_de_genes1$AveExpr,2)
top_de_genes1$t <- round(top_de_genes1$t,2)
top_de_genes1$P.Value <- signif(top_de_genes1$P.Value,3)
top_de_genes1$adj.P.Val <- signif(top_de_genes1$adj.P.Val,3)
top_de_genes1$B <- round(top_de_genes1$B,2)

# Limma topTable reduce cols
top_de_genes_small1 <- top_de_genes1[,c("SYMBOL","logFC","P.Value","adj.P.Val","B")]
str(top_de_genes1)
str(top_de_genes_small1)
top_de_genes_small1[grep("",top_de_genes_small1[,1]),]
?sort
filter(top_de_genes_small1,logFC<=-0.24)

# Save t
write.table(top_de_genes_small1, file="all_probes_limma_fdr_small_27_02_2016.csv",row.names=F,quote=FALSE,sep = ";")
write.table(top_de_genes1, file="all_probes_limma_fdr_27_02_2016.csv",row.names=F,quote=FALSE,sep = ";")

```



##Volcano Plot
```{r}
# Make a basic volcano plot
library(Tmisc)
library(calibrate)
top_de_genes_vol <- topTable(contrast.fit1, coef=1, number=5000,adjust.method="fdr",p.value=1,confint=TRUE, sort.by="logFC")

#save plot
jpeg(file = "Volcanoplot_LIMMA_For_paper_FEP.jpeg", pointsize = 20, width = 1000, height = 1000)
with(top_de_genes_vol, plot(logFC, -log10(P.Value), pch=20, main="Volcano Plot of Differentially Expressed Probes", xlim=c(-1,1.5)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(top_de_genes_vol, adj.P.Val >0.05 ), points(logFC, -log10(P.Value), pch=20, col="firebrick"))
with(subset(top_de_genes_vol, adj.P.Val <0.05 ), points(logFC, -log10(P.Value), pch=20, col="navy"))
with(subset(top_de_genes_vol, logFC <= 0.138 & logFC >= -0.152 ), points(logFC, -log10(P.Value), pch=20, col="firebrick"))


    
# Label points with the textxy function from the calibrate plot
#with(top_de_genes_vol [c(1,3,5),], textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5))
#with(top_de_genes_vol [58,], textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5,offset = 1.1))
with(subset(top_de_genes_vol, logFC >0.4 & logFC < 2 & adj.P.Val <0.05), textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5))
with(subset(top_de_genes_vol, logFC < -0.4 & adj.P.Val <0.05 ), textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5,offset = .6))
with(subset(top_de_genes_vol, adj.P.Val <0.00007), textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5))
dev.off()



str(top_de_genes_vol)
top_de_genes_vol$ProbeID [1:4]




subset(top_de_genes_vol, ProbeID  )
?subset
```




## User List Enrichement
```{r gene_list}
library(WGCNA)

all_probes <- top_de_genes_vol 
all_probes2 <-top_de_genes1

#Subset to pvalue 0.05
all_probes$Groups <-all_probes$adj.P.Val <= 0.05
all_probes$Groups<-replace(all_probes$Groups, all_probes$Groups==FALSE, "background")
all_probes$Groups<-replace(all_probes$Groups, all_probes$Groups==TRUE, "FDR_Pass")

enrichments = userListEnrichment(all_probes$TargetID,all_probes$Groups,useBloodAtlases=T,useBrainLists=T,useStemCellLists=F, useBrainRegionMarkers=F, useImmunePathwayLists=F,usePalazzoloWang=F)

enpv<-enrichments$pValue

write.csv(enpv[order(enpv$Type,enpv$Pvalues),],file="GAP_all_enrichmennts_brain_blood.csv")


#Subset to logfc 0.1
all_probes2$Groups <-all_probes2$adj.P.Val <= 0.05 & !(all_probes2$logFC <= 0.1 & all_probes2$logFC >= -0.1 )
all_probes2$Groups<-replace(all_probes2$Groups, all_probes2$Groups==FALSE, "background")
all_probes2$Groups<-replace(all_probes2$Groups, all_probes2$Groups==TRUE, "FDR_Pass")


enrichments = userListEnrichment(all_probes2$TargetID,all_probes2$Groups,useBloodAtlases=T,useBrainLists=T,useStemCellLists=F, useBrainRegionMarkers=F, useImmunePathwayLists=F,usePalazzoloWang=F)


enpv<-enrichments$pValue

write.csv(enpv[order(enpv$Type,enpv$Pvalues),],file="GAP_all_enrichmennts_with_logFC_brain_blood.csv")


```
