---
title: "LIMMA"
author: "DJL"
date: "03/11/2016"
output:
  html_document:
    toc: yes
    toc_float: yes
---
#Overview

**Aim**  
Differential Expression on GAP FEP Data.  


#Pseudocode
```{r Pseudocode, eval=FALSE}

#1.Load Libs

#2.Set Dir

#3.Load Lumibatch

#4. LIMMA
Extract pData
Prepare for Limma (age ethnicity gender Phenotype)
perform limma

#5. Write CSV files
All Probes
Significantly Diff expressed Probes
Significantly Diff expressed Probes + log2 cut off
Non Significant Probes

#6. Volcano Plot




```

## Load Libaries
```{r load_libs, tidy=TRUE}
library(lumi)
library(limma)
library(dplyr)

```

## Set directories
```{r set directories, eval=TRUE, tidy=TRUE}
data_dir <-"./data/"
P1_output_dir <-"./P1_Diff_Ex/output/"
P1_figs_dir <-"./P1_Diff_Ex/figs/"
```



## Load eset
```{r load_eset, tidy=TRUE}
#Load Gene expression data in Lumi batch object

load(paste(data_dir,"GAP_FEP_Full_Gene_Expression_Data.RData",sep=""))

```

## Limma
```{r LIMMA Design}

#Prepare covariates
names(pData(eset_bg_log2_rsn_SVA_Good))
gap <- factor(eset_bg_log2_rsn_SVA_Good$Phenotype)
gap_gender <- factor(eset_bg_log2_rsn_SVA_Good$Gender)
gap_age <-as.numeric(eset_bg_log2_rsn_SVA_Good$Age)
gap_ethnicity <-factor(eset_bg_log2_rsn_SVA_Good$Ethnicity)

#Design Matrix
design = model.matrix(~0+gap+gap_gender+gap_ethnicity+gap_age)  

## Limma lmFit
fit <- lmFit(eset_bg_log2_rsn_SVA_Good, design)

## Limma makeContrasts 
contrasts <- makeContrasts(gapFEP-gapcontrol, levels=design)

## Limma eBayes on makeContrasts
contrast.fit <- contrasts.fit(fit, contrasts)
contrast.fit <- eBayes(contrast.fit)


```

#Top Table
```{r limma_topTable, comment=NA, width=100, tidy=TRUE}
# Limma topTabl
top_de_genes <- topTable(contrast.fit1, coef=1, number=10000,adjust.method="fdr",p.value=1,confint=TRUE)

names(top_de_genes)

# tidy up
top_de_genes$logFC <- round(top_de_genes$logFC,2)
top_de_genes$CI.L <- round(top_de_genes$CI.L,2)
top_de_genes$CI.R <- round(top_de_genes$CI.R,2)
top_de_genes$AveExpr <- round(top_de_genes$AveExpr,2)
top_de_genes$t <- round(top_de_genes$t,2)
top_de_genes$P.Value <- signif(top_de_genes$P.Value,3)
top_de_genes$adj.P.Val <- signif(top_de_genes$adj.P.Val,3)
top_de_genes$B <- round(top_de_genes$B,2)


##Save full dataframe
full_limma_results_AgeSexEthnicity_Adj <- top_de_genes
write.table(full_limma_results_AgeSexEthnicity_Adj, file=paste(P1_output_dir,"full_limma_results_AgeSexEthnicity_Adj.tsv"),row.names=FALSE,quote=FALSE,sep = "\t")

##Save Reduced dataframe
small_limma_results_AgeSexEthnicity_Adj <- top_de_genes[,c("TargetID","logFC","P.Value","adj.P.Val","CHROMOSOME","DEFINITION")]
write.table(small_limma_results_AgeSexEthnicity_Adj, file=paste(P1_output_dir,"small_limma_results_AgeSexEthnicity_Adj.tsv"),row.names=FALSE,quote=FALSE,sep = "\t")


```


#### Significantly DE Probes

```{r}


de_res <- top_de_genes

de_res <- de_res %>% 
  mutate(SIG_DE=adj.P.Val <=0.05, 
         LogFC_DIRECTION=ifelse(logFC >= 0, "up-regulated",
                          ifelse(logFC < 0, "down-regulated", "no-change")),
         LogFC_BIOLOCICAL=ifelse(logFC >= 0.1, "up-regulated",
                         ifelse(logFC <= -0.1, "down-regulated", "no-sig-change")),
         PROBE_KEEP=ifelse( grepl("^LOC",TargetID),"DROP",
                       ifelse( grepl("^HS\\.",TargetID), "DROP","KEEP"))
         )
      
## Significantly DE Probes
datatoplot<- de_res %>% filter(PROBE_KEEP=="KEEP")

## Number of probes
table(datatoplot$PROBE_KEEP)

## Number of significant probes
table(datatoplot$adj.P.Val <= 0.05)

## Number if sig probes using B-statistic
table(datatoplot$B > 0)

##All probes
write.table(de_res, file=paste(P1_output_dir,"Supplementary_Table_1_all_probes_5797_limma_results_AgeSexEthnicity_Adj.tsv",sep=""),row.names=FALSE,quote=FALSE,sep = "\t")

##Significant probes
datatoplot_sig<-datatoplot[datatoplot$adj.P.Val <= 0.05,]
dim(datatoplot_sig)
write.table(datatoplot_sig, file=paste(P1_output_dir,"Supplementary_Table_2_significant_limma_results_AgeSexEthnicity_Adj.tsv",sep=""),row.names=FALSE,quote=FALSE,sep = "\t")

##upregulated
datatoplot_sig_up<-datatoplot_sig[datatoplot_sig$logFC > 0,]
dim(datatoplot_sig_up)
write.table(datatoplot_sig_up, file=paste(P1_output_dir,"Supplementary_Table_3_upregulated_significant_limma_results_AgeSexEthnicity_Adj.tsv",sep=""),row.names=FALSE,quote=FALSE,sep = "\t")


table1up<-datatoplot_sig_up[1:50,c("TargetID","CHROMOSOME","DEFINITION","logFC","adj.P.Val")]

table1up$logFC<-signif(table1up$logFC,2)
table1up$adj.P.Val<-signif(table1up$adj.P.Val,2)

write.table(table1up, file=paste(P1_output_dir,"Table_1_upregulated_significant_limma_results_AgeSexEthnicity_Adj.tsv",sep=""),row.names=FALSE,quote=FALSE,sep = "\t")


##downregulated
datatoplot_sig_down<-datatoplot_sig[datatoplot_sig$logFC < 0,]
dim(datatoplot_sig_down)
write.table(datatoplot_sig_down, file=paste(P1_output_dir,"Supplementary_Table_5_downregulated_significant_limma_results_AgeSexEthnicity_Adj.tsv",sep=""),row.names=FALSE,quote=FALSE,sep = "\t")

table2down<-datatoplot_sig_down[1:50,c("TargetID","CHROMOSOME","DEFINITION","logFC","adj.P.Val")]

table2down$logFC<-signif(table2down$logFC,2)
table2down$adj.P.Val<-signif(table2down$adj.P.Val,2)

write.table(table2down, file=paste(P1_output_dir,"Table_2_downregulated_significant_limma_results_AgeSexEthnicity_Adj.tsv",sep=""),row.names=FALSE,quote=FALSE,sep = "\t")

```


##Volcano Plot
```{r}
# Make a basic volcano plot
library(Tmisc)
library(calibrate)

filename <- "Volcanoplot.jpeg"
jpeg(file = paste(P1_figs_dir,filename,sep=""), pointsize = 20, width = 1500, height = 1300)
with(datatoplot, plot(logFC, -log10(P.Value), pch=20, main="Volcano Plot of Differentially Expressed Probes", xlim=c(-0.5,1.5)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(datatoplot, logFC >= 0 ), points(logFC, -log10(P.Value), pch=20, col="navy"))
with(subset(datatoplot, logFC <= 0 ), points(logFC, -log10(P.Value), pch=20, col="darkgreen"))
with(subset(datatoplot, adj.P.Val >0.05 ), points(logFC, -log10(P.Value), pch=20, col="firebrick"))
    
# Label points with the textxy function from the calibrate plot
with(datatoplot [c(6:10),], textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5))
with(subset(datatoplot, logFC < -0.27 & adj.P.Val <0.05 ), textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5,offset = .6))
with(subset(datatoplot, logFC >= 0.37 & adj.P.Val <0.05 ), textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5,offset = .6))
dev.off()


filename <- "Volcanoplot.pdf"
pdf(paste(P1_figs_dir,filename,sep=""), width = 15, height = 15,pointsize=10)
with(datatoplot, plot(logFC, -log10(P.Value), pch=20, main="Volcano Plot of Differentially Expressed Probes", xlim=c(-0.5,1.5)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(datatoplot, logFC >= 0 ), points(logFC, -log10(P.Value), pch=20, col="navy"))
with(subset(datatoplot, logFC <= 0 ), points(logFC, -log10(P.Value), pch=20, col="darkgreen"))
with(subset(datatoplot, adj.P.Val >0.05 ), points(logFC, -log10(P.Value), pch=20, col="firebrick"))
    
# Label points with the textxy function from the calibrate plot
with(datatoplot [c(6:10),], textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5))
with(subset(datatoplot, logFC < -0.27 & adj.P.Val <0.05 ), textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5,offset = .6))
with(subset(datatoplot, logFC >= 0.37 & adj.P.Val <0.05 ), textxy(logFC, -log10(P.Value), labs=TargetID, cex=.5,offset = .6))
dev.off()


#LogFC by chromosome
title = "LogFC by Chromosome"
ggplot(data = datatoplot, 
      aes(x = CHROMOSOME, y=logFC, color=CHROMOSOME)) +
      geom_boxplot(alpha = 0)+
      geom_text(data=filter(datatoplot, logFC >= 0.3 |logFC <= -0.27),check_overlap = TRUE,angle=45,
            aes(CHROMOSOME,logFC,label=SYMBOL))+
      ggtitle(title)
ggsave(paste(P1_figs_dir,title,".png",sep=""))

table(datatoplot$CHROMOSOME)



datatoplot %>% filter(CHROMOSOME == c(21)) $>$ PROBE_COORDINATES

```



